File: moui_shell/test_probe/android/src/dev/wzzc/moui/shell/testprobe/MoUIShellTestProbePlugin.kt
package dev.wzzc.moui.shell.testprobe

import android.content.ClipData
import android.content.ClipboardManager
import android.content.Context
import android.graphics.Color
import android.graphics.RectF
import android.util.Log
import android.view.View
import android.widget.TextView
import dev.wzzc.moui.shell.MoUIHostServiceCompletion
import dev.wzzc.moui.shell.MoUIHostServiceHandler
import dev.wzzc.moui.shell.MoUIHostServiceRequest
import dev.wzzc.moui.shell.MoUIHostServiceTask
import dev.wzzc.moui.shell.MoUIHostServices
import dev.wzzc.moui.shell.MoUIShellPlugin
import dev.wzzc.moui.shell.MoUIShellPluginCapabilities
import dev.wzzc.moui.shell.MoUIPluginSubscription
import dev.wzzc.moui.shell.MoUIPlatformViewEventSink
import dev.wzzc.moui.shell.MoUIPlatformViewFactory
import dev.wzzc.moui.shell.MoUIPlatformViewPlacement
import dev.wzzc.moui.shell.MoUIPlatformViews
import org.json.JSONObject
import java.util.concurrent.atomic.AtomicBoolean
import java.util.concurrent.atomic.AtomicInteger

private const val PLUGIN_ID = "dev.wzzc.moui.shell.test-probe"
private const val PLATFORM_VIEW_KIND = "$PLUGIN_ID.view"
private const val HOST_CHANNEL = "$PLUGIN_ID.channel"
private const val TEST_PROBE_GATE = "moui.shell.testProbe"
private const val SERVICE_TEXT_LABEL = "Service probe text"
private const val SERVICE_ACTION_LABEL = "Activate service probe"
private const val SERVICE_PROBE_TEXT = "ime-shell-probe"
private const val LOG_TAG = "MoUIShell"

private object ProbeState {
    val platformViewCreate = AtomicInteger(0)
    val platformViewResize = AtomicInteger(0)
    val platformViewClip = AtomicInteger(0)
    val platformViewEvent = AtomicInteger(0)
    val platformViewDispose = AtomicInteger(0)
    val hostChannelSuccess = AtomicInteger(0)
    val hostChannelError = AtomicInteger(0)
    val hostChannelCancel = AtomicInteger(0)
    val hostChannelExactlyOnce = AtomicInteger(0)
    val hostChannelLateAfterDispose = AtomicInteger(0)
    val serviceSmokeFired = AtomicInteger(0)
    val serviceSmokeCompleted = AtomicInteger(0)

    fun increment(counter: AtomicInteger) {
        counter.incrementAndGet()
        Log.i(LOG_TAG, "moui-shell test-probe snapshot=${snapshot()}")
    }

    fun snapshot(): String = JSONObject()
        .put("platformViewCreate", platformViewCreate.get())
        .put("platformViewResize", platformViewResize.get())
        .put("platformViewClip", platformViewClip.get())
        .put("platformViewEvent", platformViewEvent.get())
        .put("platformViewDispose", platformViewDispose.get())
        .put("hostChannelSuccess", hostChannelSuccess.get())
        .put("hostChannelError", hostChannelError.get())
        .put("hostChannelCancel", hostChannelCancel.get())
        .put("hostChannelExactlyOnce", hostChannelExactlyOnce.get())
        .put("hostChannelLateAfterDispose", hostChannelLateAfterDispose.get())
        .put("serviceSmokeFired", serviceSmokeFired.get())
        .put("serviceSmokeCompleted", serviceSmokeCompleted.get())
        .toString()
}

private object ProbeServiceSmoke {
    private val installed = AtomicBoolean(false)
    private val fired = AtomicBoolean(false)
    private var subscription: MoUIPluginSubscription? = null

    fun install(context: Context, capabilities: MoUIShellPluginCapabilities) {
        if (!installed.compareAndSet(false, true)) return
        val applicationContext = context.applicationContext
        subscription = capabilities.semantics.observe { snapshot, runtimeInput ->
            if (!capabilities.launchOptions.isEnabled(TEST_PROBE_GATE) || fired.get()) {
                return@observe
            }
            val textField = snapshot.nodes.firstOrNull {
                it.label == SERVICE_TEXT_LABEL && it.role == "TextField"
            } ?: return@observe
            val action = snapshot.nodes.firstOrNull {
                it.label == SERVICE_ACTION_LABEL && it.role == "Button"
            } ?: return@observe
            if (!fired.compareAndSet(false, true)) return@observe
            ProbeState.increment(ProbeState.serviceSmokeFired)
            Log.i(LOG_TAG, "moui-shell service smoke begin")
            val textFocusAccepted = runtimeInput.dispatchAccessibility(textField.elementId, 1)
            var accepted = textFocusAccepted
            val setTextAccepted = runtimeInput.dispatchAccessibility(
                textField.elementId,
                2,
                SERVICE_PROBE_TEXT,
            )
            accepted = setTextAccepted && accepted
            val imeCommitAccepted = runtimeInput.dispatchTextInput(1, SERVICE_PROBE_TEXT, 0, 0)
            accepted = imeCommitAccepted && accepted
            if (imeCommitAccepted) Log.i(LOG_TAG, "moui-shell service ime edit kind=commit")
            val selectionAccepted = runtimeInput.dispatchTextInput(
                2,
                "",
                0,
                SERVICE_PROBE_TEXT.length,
            )
            accepted = selectionAccepted && accepted
            val copyAccepted = runtimeInput.dispatchCommand(0)
            accepted = copyAccepted && accepted
            if (copyAccepted) Log.i(LOG_TAG, "moui-shell service smoke copy")
            val clipboard = applicationContext.getSystemService(ClipboardManager::class.java)
            clipboard?.setPrimaryClip(ClipData.newPlainText("MoUI", "clipboard-service-probe"))
            accepted = clipboard != null && accepted
            val pasteAccepted = runtimeInput.dispatchCommand(2)
            accepted = pasteAccepted && accepted
            if (pasteAccepted) Log.i(LOG_TAG, "moui-shell service smoke paste")
            val cutAccepted = runtimeInput.dispatchCommand(1)
            accepted = cutAccepted && accepted
            if (cutAccepted) Log.i(LOG_TAG, "moui-shell service smoke cut")
            val actionFocusAccepted = runtimeInput.dispatchAccessibility(action.elementId, 1)
            accepted = actionFocusAccepted && accepted
            if (actionFocusAccepted) {
                Log.i(LOG_TAG, "moui-shell service accessibility focus id=${action.elementId}")
            }
            val actionActivateAccepted = runtimeInput.dispatchAccessibility(action.elementId, 0)
            accepted = actionActivateAccepted && accepted
            if (actionActivateAccepted) {
                Log.i(LOG_TAG, "moui-shell service accessibility action=activate id=${action.elementId}")
            }
            if (accepted && runtimeInput.isActive) {
                ProbeState.increment(ProbeState.serviceSmokeCompleted)
            }
            Log.i(LOG_TAG, "moui-shell service smoke end accepted=${if (accepted) 1 else 0}")
        }
    }
}

private class ProbeTextView(
    context: Context,
    val events: MoUIPlatformViewEventSink,
) : TextView(context) {
    var lastFrame: RectF? = null
    var sentReady = false

    init {
        text = "MoUI test probe"
        setTextColor(Color.WHITE)
        setBackgroundColor(Color.rgb(22, 101, 52))
        contentDescription = "MoUI test probe PlatformView"
        setOnClickListener {
            if (events.dispatch("activate", "android", ProbeState.snapshot(), true)) {
                ProbeState.increment(ProbeState.platformViewEvent)
            }
        }
    }
}

private object ProbePlatformViewFactory : MoUIPlatformViewFactory {
    override fun create(context: Context, id: String, sink: MoUIPlatformViewEventSink): View {
        ProbeState.increment(ProbeState.platformViewCreate)
        return ProbeTextView(context, sink).apply { tag = id }
    }

    override fun update(view: View, placement: MoUIPlatformViewPlacement) {
        val probe = view as ProbeTextView
        val previous = probe.lastFrame
        if (previous == null || previous.width() != placement.frame.width() ||
            previous.height() != placement.frame.height()
        ) {
            ProbeState.increment(ProbeState.platformViewResize)
        }
        if (placement.clip != null) ProbeState.increment(ProbeState.platformViewClip)
        probe.lastFrame = RectF(placement.frame)
        probe.text = placement.properties["label"] ?: "MoUI test probe"
        if (!probe.sentReady) {
            probe.sentReady = true
            if (probe.events.dispatch("ready", placement.id, ProbeState.snapshot(), true)) {
                ProbeState.increment(ProbeState.platformViewEvent)
            }
        }
    }

    override fun dispose(view: View) {
        val probe = view as ProbeTextView
        ProbeState.increment(ProbeState.platformViewDispose)
        probe.events.dispatch("disposed", placementValue(view), ProbeState.snapshot(), false)
        probe.setOnClickListener(null)
    }

    private fun placementValue(view: View): String = view.tag?.toString() ?: ""
}

private class ProbePendingTask(
    private val completion: MoUIHostServiceCompletion,
    private val kind: Kind,
) : MoUIHostServiceTask {
    private val cancelled = AtomicBoolean(false)

    enum class Kind {
        CANCEL,
        LATE_AFTER_DISPOSE,
    }

    override fun cancel() {
        if (!cancelled.compareAndSet(false, true)) return
        when (kind) {
            Kind.CANCEL -> ProbeState.increment(ProbeState.hostChannelCancel)
            Kind.LATE_AFTER_DISPOSE -> if (!completion.ok("late-after-dispose")) {
                ProbeState.increment(ProbeState.hostChannelLateAfterDispose)
            }
        }
    }
}

private object ProbeHostServiceHandler : MoUIHostServiceHandler {
    override fun handle(
        request: MoUIHostServiceRequest,
        completion: MoUIHostServiceCompletion,
    ): MoUIHostServiceTask? = when (request.operation) {
        "success", "echo" -> {
            if (completion.ok(request.payload)) {
                ProbeState.increment(ProbeState.hostChannelSuccess)
            }
            null
        }
        "snapshot" -> {
            completion.ok(ProbeState.snapshot())
            null
        }
        "error" -> {
            if (completion.error(request.payload.ifEmpty { "test-probe error" })) {
                ProbeState.increment(ProbeState.hostChannelError)
            }
            null
        }
        "exactly-once" -> {
            val firstAccepted = completion.ok("first")
            val duplicateAccepted = completion.error("duplicate")
            if (firstAccepted && !duplicateAccepted) {
                ProbeState.increment(ProbeState.hostChannelExactlyOnce)
            }
            null
        }
        "cancel" -> ProbePendingTask(completion, ProbePendingTask.Kind.CANCEL)
        "late-after-dispose" -> ProbePendingTask(completion, ProbePendingTask.Kind.LATE_AFTER_DISPOSE)
        else -> {
            completion.unavailable("unknown test-probe operation: ${request.operation}")
            null
        }
    }
}

class MoUIShellTestProbePlugin : MoUIShellPlugin {
    override val id: String = PLUGIN_ID

    override fun install(context: Context) {
        context.applicationContext
        MoUIPlatformViews.register(PLATFORM_VIEW_KIND, ProbePlatformViewFactory)
        MoUIHostServices.register(HOST_CHANNEL, ProbeHostServiceHandler)
    }

    override fun install(context: Context, capabilities: MoUIShellPluginCapabilities) {
        install(context)
        ProbeServiceSmoke.install(context, capabilities)
    }
}

File: moui_shell/test_probe/ios/src/MoUIShellTestProbePlugin.swift
import Foundation
import MoUIShellShell
import UIKit

private let pluginIdentifier = "dev.wzzc.moui.shell.test-probe"
private let platformViewKind = "\(pluginIdentifier).view"
private let hostChannel = "\(pluginIdentifier).channel"
private let testProbeGate = "moui.shell.testProbe"
private let serviceTextLabel = "Service probe text"
private let serviceActionLabel = "Activate service probe"
private let serviceProbeText = "ime-shell-probe"

private enum ProbeCounter: String, CaseIterable {
  case platformViewCreate
  case platformViewResize
  case platformViewClip
  case platformViewEvent
  case platformViewDispose
  case hostChannelSuccess
  case hostChannelError
  case hostChannelCancel
  case hostChannelExactlyOnce
  case hostChannelLateAfterDispose
  case serviceSmokeFired
  case serviceSmokeCompleted
}

private final class ProbeState {
  static let shared = ProbeState()

  private let lock = NSLock()
  private var counters = Dictionary(
    uniqueKeysWithValues: ProbeCounter.allCases.map { ($0.rawValue, 0) }
  )

  func increment(_ counter: ProbeCounter) {
    lock.lock()
    counters[counter.rawValue, default: 0] += 1
    let value = counters
    lock.unlock()
    NSLog("moui-shell test-probe snapshot=%@", Self.encode(value))
  }

  func snapshot() -> String {
    lock.lock()
    let value = counters
    lock.unlock()
    return Self.encode(value)
  }

  private static func encode(_ value: [String: Int]) -> String {
    guard
      let data = try? JSONSerialization.data(withJSONObject: value, options: [.sortedKeys]),
      let text = String(data: data, encoding: .utf8)
    else { return "{}" }
    return text
  }
}

private final class ProbePlatformView: UILabel {
  var eventSink: MOUIShellPlatformViewEventSink?
  var lastSize: CGSize?
  var sentReady = false

  override init(frame: CGRect) {
    super.init(frame: frame)
    backgroundColor = UIColor(red: 0.09, green: 0.40, blue: 0.20, alpha: 1)
    textColor = .white
    textAlignment = .center
    text = "MoUI test probe"
    isUserInteractionEnabled = true
    accessibilityIdentifier = "moui-shell-test-probe-platform-view"
    addGestureRecognizer(UITapGestureRecognizer(target: self, action: #selector(activate)))
  }

  @available(*, unavailable)
  required init?(coder: NSCoder) {
    fatalError("init(coder:) is unavailable")
  }

  @objc private func activate() {
    if eventSink?.send(
      name: "activate",
      value: "ios",
      detail: ProbeState.shared.snapshot(),
      flag: true
    ) == true {
      ProbeState.shared.increment(.platformViewEvent)
    }
  }
}

private final class ProbePlatformViewFactory: MOUIShellPlatformViewFactory {
  func makePlatformView(identifier: String) -> UIView {
    ProbeState.shared.increment(.platformViewCreate)
    let view = ProbePlatformView(frame: .zero)
    view.accessibilityValue = identifier
    return view
  }

  func updatePlatformView(
    _ view: UIView,
    placement: MOUIShellPlatformViewPlacement,
    events: MOUIShellPlatformViewEventSink
  ) {
    guard let probe = view as? ProbePlatformView else { return }
    if probe.lastSize != placement.frame.size {
      ProbeState.shared.increment(.platformViewResize)
      probe.lastSize = placement.frame.size
    }
    if placement.clip != nil { ProbeState.shared.increment(.platformViewClip) }
    probe.eventSink = events
    probe.text = placement.properties["label"] ?? "MoUI test probe"
    if !probe.sentReady {
      probe.sentReady = true
      if events.send(
        name: "ready",
        value: placement.identifier,
        detail: ProbeState.shared.snapshot(),
        flag: true
      ) {
        ProbeState.shared.increment(.platformViewEvent)
      }
    }
  }

  func disposePlatformView(_ view: UIView) {
    guard let probe = view as? ProbePlatformView else { return }
    ProbeState.shared.increment(.platformViewDispose)
    probe.eventSink?.send(
      name: "disposed",
      value: probe.accessibilityValue ?? "",
      detail: ProbeState.shared.snapshot()
    )
    probe.eventSink = nil
    probe.gestureRecognizers?.forEach { probe.removeGestureRecognizer($0) }
  }
}

private final class ProbePendingTask: MOUIEmbedderHostChannelTask {
  enum Kind {
    case cancel
    case lateAfterDispose
  }

  private let lock = NSLock()
  private var cancelled = false
  private let completion: MOUIEmbedderHostChannelCompletion
  private let kind: Kind

  init(completion: MOUIEmbedderHostChannelCompletion, kind: Kind) {
    self.completion = completion
    self.kind = kind
  }

  func cancel() {
    lock.lock()
    guard !cancelled else {
      lock.unlock()
      return
    }
    cancelled = true
    lock.unlock()
    switch kind {
    case .cancel:
      ProbeState.shared.increment(.hostChannelCancel)
    case .lateAfterDispose:
      if !completion.succeed(payload: "late-after-dispose") {
        ProbeState.shared.increment(.hostChannelLateAfterDispose)
      }
    }
  }
}

private final class ProbeHostChannelHandler: MOUIEmbedderHostChannelHandler {
  func handle(
    request: MOUIEmbedderHostChannelRequest,
    completion: MOUIEmbedderHostChannelCompletion
  ) -> MOUIEmbedderHostChannelTask? {
    switch request.operation {
    case "success", "echo":
      if completion.succeed(payload: request.payload) {
        ProbeState.shared.increment(.hostChannelSuccess)
      }
      return nil
    case "snapshot":
      completion.succeed(payload: ProbeState.shared.snapshot())
      return nil
    case "error":
      if completion.fail(payload: request.payload.isEmpty ? "test-probe error" : request.payload) {
        ProbeState.shared.increment(.hostChannelError)
      }
      return nil
    case "exactly-once":
      let firstAccepted = completion.succeed(payload: "first")
      let duplicateAccepted = completion.fail(payload: "duplicate")
      if firstAccepted && !duplicateAccepted {
        ProbeState.shared.increment(.hostChannelExactlyOnce)
      }
      return nil
    case "cancel":
      return ProbePendingTask(completion: completion, kind: .cancel)
    case "late-after-dispose":
      return ProbePendingTask(completion: completion, kind: .lateAfterDispose)
    default:
      completion.unavailable(payload: "unknown test-probe operation: \(request.operation)")
      return nil
    }
  }
}

private final class ProbeServiceSmoke {
  static let shared = ProbeServiceSmoke()

  private var fired = false
  private var subscription: MOUIShellPluginSubscription?

  func install(capabilities: MOUIShellPluginCapabilities) {
    guard subscription == nil else { return }
    subscription = capabilities.semantics.observe { [weak self, weak capabilities] snapshot, runtimeInput in
      guard
        let self,
        let capabilities,
        capabilities.launchOptions.isEnabled(testProbeGate),
        !self.fired,
        let textField = snapshot.nodes.first(where: {
          $0.label == serviceTextLabel && $0.role == "TextField"
        }),
        let action = snapshot.nodes.first(where: {
          $0.label == serviceActionLabel && $0.role == "Button"
        })
      else { return }
      self.fired = true
      ProbeState.shared.increment(.serviceSmokeFired)
      NSLog("moui-shell service smoke begin")
      let textFocusAccepted = runtimeInput.dispatchAccessibility(
        elementIdentifier: textField.elementIdentifier,
        action: 1
      )
      var accepted = textFocusAccepted
      let setTextAccepted = runtimeInput.dispatchAccessibility(
        elementIdentifier: textField.elementIdentifier,
        action: 2,
        value: serviceProbeText
      )
      accepted = setTextAccepted && accepted
      let imeCommitAccepted = runtimeInput.dispatchTextInput(
        kind: 1,
        text: serviceProbeText,
        start: 0,
        end: 0
      )
      accepted = imeCommitAccepted && accepted
      if imeCommitAccepted { NSLog("moui-shell service ime edit kind=commit") }
      let selectionAccepted = runtimeInput.dispatchTextInput(
        kind: 2,
        text: "",
        start: 0,
        end: Int32(serviceProbeText.utf16.count)
      )
      accepted = selectionAccepted && accepted
      let copyAccepted = runtimeInput.dispatchCommand(kind: 0)
      accepted = copyAccepted && accepted
      if copyAccepted { NSLog("moui-shell service smoke copy") }
      UIPasteboard.general.string = "clipboard-service-probe"
      let pasteAccepted = runtimeInput.dispatchCommand(kind: 2)
      accepted = pasteAccepted && accepted
      if pasteAccepted { NSLog("moui-shell service smoke paste") }
      let cutAccepted = runtimeInput.dispatchCommand(kind: 1)
      accepted = cutAccepted && accepted
      if cutAccepted { NSLog("moui-shell service smoke cut") }
      let actionFocusAccepted = runtimeInput.dispatchAccessibility(
        elementIdentifier: action.elementIdentifier,
        action: 1
      )
      accepted = actionFocusAccepted && accepted
      if actionFocusAccepted {
        NSLog("moui-shell service accessibility focus id=%d", action.elementIdentifier)
      }
      let actionActivateAccepted = runtimeInput.dispatchAccessibility(
        elementIdentifier: action.elementIdentifier,
        action: 0
      )
      accepted = actionActivateAccepted && accepted
      if actionActivateAccepted {
        NSLog("moui-shell service accessibility action=activate id=%d", action.elementIdentifier)
      }
      if accepted && runtimeInput.isActive {
        ProbeState.shared.increment(.serviceSmokeCompleted)
      }
      NSLog("moui-shell service smoke end accepted=%d", accepted ? 1 : 0)
    }
  }
}

public enum MoUIShellTestProbePlugin: MOUIShellPlugin {
  public static let identifier = pluginIdentifier

  public static func install(in registry: MOUIShellPluginRegistry) {
    registry.registerPlatformView(kind: platformViewKind, factory: ProbePlatformViewFactory())
    registry.registerHostChannel(name: hostChannel, handler: ProbeHostChannelHandler())
  }

  public static func install(
    in registry: MOUIShellPluginRegistry,
    capabilities: MOUIShellPluginCapabilities
  ) {
    install(in: registry)
    ProbeServiceSmoke.shared.install(capabilities: capabilities)
  }
}
